http: add IPv6 range support for NO_PROXY environment variable

haramj opened the pull request:
This PR resolves the `// TODO(joyeecheung): support IPv6.` comment in `internal/http.js` by adding comprehensive support for IPv6 address ranges to the `NO_PROXY` environment variable.

Previously, the `ProxyConfig` class's `shouldUseProxy` method could only handle IPv4 address ranges. With this change, users can now specify IPv6 ranges (e.g., `::1-::100`) to correctly bypass proxies for hosts within those ranges.

The implementation involves:
* A new helper function `ipv6ToBigInt` to convert 128-bit IPv6 addresses into `BigInt`s for reliable numerical comparison.
* Updated `shouldUseProxy` logic to check if a host is within a given IPv6 range.

This update ensures the proxy bypass feature is robust and consistent with both IPv4 and IPv6 network standards.

New test cases have been added to `test/client-proxy/test-http-proxy-request-no-proxy-ipv6.mjs` to verify the new functionality.

Test
* A host within the specified IPv6 range (`::1-::100`) correctly bypasses the proxy.
* A host outside the specified IPv6 range correctly uses the proxy.

This ensures that the new feature works as expected and does not introduce regressions to existing functionality.

/cc @joyeecheung 
<!--
Before submitting a pull request, please read:

- the CONTRIBUTING guide at https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md
- the commit message formatting guidelines at
  https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

If you believe this PR should be highlighted in the Node.js CHANGELOG
please add the `notable-change` label.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

haramj committed:
http: add IPv6 range support for NO_PROXY environment variable

Signed-off-by: haramjeong <04harams77@gmail.com>

haramj committed:
http: remove TODO and add condition

Signed-off-by: haramjeong <04harams77@gmail.com>

trivikr commented:
@haramj Can you please rebase, sign the first commit and force push?

It's required as per updated [instructions](https://github.com/nodejs/node/blob/main/doc/contributing/pull-requests.md#step-4-commit)
> Your commit must contain the Signed-off-by line with your name and email address as an acknowledgement that you agree to the [Developer Certificate of Origin](https://github.com/nodejs/node/blob/main/CONTRIBUTING.md#developers-certificate-of-origin-11)

haramj commented:
> @haramj Can you please rebase, sign the first commit and force push?
> 
> It's required as per updated [instructions](https://github.com/nodejs/node/blob/main/doc/contributing/pull-requests.md#step-4-commit)
> 
> > Your commit must contain the Signed-off-by line with your name and email address as an acknowledgement that you agree to the [Developer Certificate of Origin](https://github.com/nodejs/node/blob/main/CONTRIBUTING.md#developers-certificate-of-origin-11)

Thanks! I’ve rebased the branch onto the latest main and added the required Signed-off-by line to both commits.

haramj commented:
@trivikr It looks like the failing test is unrelated to this change (parallel/test-repl-use-global failed with ERR_INSPECTOR_COMMAND in the BoringSSL shared-library job). Could someone please rerun the failed job?

jasnell commented on lib/internal/http.js:
```diff
@@ -70,6 +72,35 @@ function ipToInt(ip) {
+
+  if (parts.length === 2) {
+    const leftParts = left.split(':').filter((p) => p.length > 0);
+    const rightParts = right.split(':').filter((p) => p.length > 0);
```
Nit: just eliminate the extra closure.

```suggestion
    const filter = (p) => p.length > 0;
    const leftParts = left.split(':').filter(filter);
    const rightParts = right.split(':').filter(filter);
```

haramj replied:
This code was removed as part of switching the IPv6 range check to BlockList.

jasnell commented on lib/internal/http.js:
```diff
@@ -70,6 +72,35 @@ function ipToInt(ip) {
   return result >>> 0;
 }
 
+function ipv6ToBigInt(ip) {
```
We already have `net.BlockList` implemented and available in the runtime. We should just use it.

```js
const bl = new BlockList();
bl.addRange('::0', '::100', 'ipv6');
console.log(bl.check('::1', 'ipv6');  // true
console.log(bl.check('::101', 'ipv6');  // false
```

haramj replied:
Updated to use BlockList for IPv6 range checks and removed the custom IPv6-to-BigInt conversion logic. Thanks!

haramj committed:
http: use BlockList for IPv6 proxy bypass ranges

haramj commented:
@jasnell Thanks for the review. I updated the implementation to use BlockList for IPv6 proxy bypass ranges and all CI checks are now passing.

Could you please take another look when you have a chance?
